refcount.h

```diff
@@ -0,0 +1,111 @@
+ * itself, and then call janus_refcall_init() to set it up. Initializing
+ * the reference counter just needs a pointer to the function to invoke
+ * when the object needs to be destroyed (counter reaches 0), while it
+ * will automatically set the counter to 1. To increase and decrease the
+ * counter just call janus_refcount_increase() and janus_refcount_decrease().
+ * When the counter reaches 0, the function passed when initializing it will
+ * be invoked: this means it's up to you to then free all the resources
+ * the object may have allocated. Notice that if this involves other
+ * objects that are reference counted, freeing the resource will just
+ * mean decreasing the related counter, and not destroying it right away.
+ *
+ * The free function must be defined like this:
+ *
+\verbatim
+void my_free_function(janus_refcount *counter);
+\endverbatim
+ *
+ * Since the reference counter cannot know the size of the object to be
+ * freed, or where in the list of members the counter has been placed,
+ * retrieving the pointer to the object to free is up to you, using the
+ * janus_refcount_containerof macro. This is an example of how the
+ * free function we have defined above may be implemented:
+ * 
+\verbatim
+typedef my_struct {
+	int number;
+	char *string;
+	janus_refcount myref;
+}
+
+void my_free_function(janus_refcount *counter) {
+	struct my_struct *my_object = janus_refcount_containerof(counter, struct my_struct, myref);
+	if(my_object->string)
+		free(my_object->string);
+	free(my_object);
+}
+\endverbatim
+ *
+ * \ingroup core
+ * \ref core
+ */
+ 
+#ifndef _JANUS_REFCOUNT_H
+#define _JANUS_REFCOUNT_H
+
+#include <glib.h>
+
+
+/*! \brief Macro to programmatically address the object itself from its counter
+ * \details \c refptr is the pointer to the janus_refcount instance, \c type
+ * is the type of the object itself (e.g., <code>struct mystruct</code>),
+ * while \c member is how the janus_refcount instance is called in the
+ * object that contains it. */
+#define janus_refcount_containerof(refptr, type, member) \
+	((type *)((char *)(refptr) - offsetof(type, member)))
+
+
+/*! \brief Janus reference counter structure */ 
+typedef struct janus_refcount janus_refcount; 
+struct janus_refcount {
+	/*! \brief The reference counter itself */
+	gint count;
+	/*! \brief Pointer to the function that will be used to free the object */
+	void (*free)(const janus_refcount *);
+};
+
+
+/*! \brief Janus reference counter initialization
+ * \note Also sets the counter to 1 automatically, so no need to increase
+ * it again manually via janus_refcount_increase() after the initialization
+ * @param ref Pointer to the Janus reference counter instance
+ * @param free Pointer to the function to invoke when the object the counter
+ * refers to needs to be destroyed */ 
+static inline void janus_refcount_init(janus_refcount *ref, void (*free)(const janus_refcount *)) {
+	/* Initialize the reference counter */
+	ref->count = 1;
+	ref->free = free;
```

Comment:
it's a tiny bit awkward that, in this function, `free` locally overrides the symbol for the libc `free()`. I might the argument `free_fn`